[RFC] Introduce and document new release process (#3793)
bpftrace/bpftrace: docs/release_process.md

We choose to follow semantic versioning. Note that this doesn't matter much for
major version < 1 but will matter a lot for >= 1.0.0 releases.

See https://semver.org/.

> ## Release schedule

Author: Makes sense, will do that.

bpftrace is released twice a year. Since our biggest dependency which also tends
to break things is LLVM, we align with the [LLVM release
schedule](https://llvm.org/docs/HowToReleaseLLVM.html). In particular, a
bpftrace release should happen **two weeks after a major LLVM release**.

In addition, four weeks before the bpftrace release, we create a stabilization
release branch which will only receive bug fixes from that moment.

> The overview of the release schedule is:

Reviewer: ```suggestion
Overview of the release cadence:
```

| bpftrace release       | **2 weeks after the LLVM release**  | [Tagging a release](#tagging-a-release)                              |

## Creating a release branch

A release branch should be created one month before the planned bpftrace
release. From that moment, only relevant bug fixes should be backported to the
branch. Similarly, after the release is done, backport PRs should be opened
against this branch.

The purpose of the branch is to give sufficient time to test features in the
upcoming rebase without blocking the development on the master branch.

When creating a branch, the following steps (submit PR where applicable) should
be performed:
1. Create a new branch according to the [Branching model](#branching-model).
1. Update Nixpks to the latest version to get the latest (pre-release) LLVM by
   running
   ```
   nix flake update
   ```
   and committing the `flake.lock` changes to the repo.
1. Bump the supported LLVM version in `CMakeLists.txt` and `flake.nix`, resolve
   any potential issues, and add a CI job to `.github/workflows/ci.yml` for the

>    new version.

Reviewer 2: These steps happen ~2 weeks after the branch is cut, right? The instructions make it sound like it should happen when you cut the release branch.

Reviewer: That's not how I read it. I read it as this happens (meaning the creation of the release branch) 4 weeks before the actual release, which is 2 weeks before the LLVM release.

Author: At the time of the branch cut, unstable Nix will have LLVM .rc3 which will be pretty similar to the final release so this step can happen immediately after the branch cut to start testing early.

OTOH, it makes sense to update to released LLVM once it's out so let me add a new step to do that after the LLVM release.

We choose to follow semantic versioning. Note that this doesn't matter much for
major version < 1 but will matter a lot for >= 1.0.0 releases.

See https://semver.org/.

## Release cadence

bpftrace is released twice a year. Since our biggest dependency, which also
tends to break things, is LLVM, we align with the [LLVM release
schedule](https://llvm.org/docs/HowToReleaseLLVM.html). In particular, a
bpftrace release should happen **two weeks after a major LLVM release**.

In addition, four weeks before the bpftrace release, we create a stabilized

> release branch, which will only receive bug fixes from that moment.

Reviewer: nit
```suggestion
release branch, which will only receive bug fixes affecting the release branch.
```

Author: Hm, that sounds a bit weird to me, too much repetition of "release branch". How about
```suggestion
release branch, which will only receive bug fixes affecting the release itself.
```

> Overview of the release cadence is:

Reviewer: ```suggestion
Overview of the release cadence is as follows:
```

| Task                   | Approximate date                    | Details                                                              |
| ---------------------- | ----------------------------------- | -------------------------------------------------------------------- |
| release branch created | **2 weeks before the LLVM release** | [Creating a release branch](#creating-a-release-branch)              |
| LLVM release           | usually second Tuesday of Mar/Sep   | [LLVM release schedule](https://llvm.org/docs/HowToReleaseLLVM.html) |
| bpftrace release       | **2 weeks after the LLVM release**  | [Tagging a release](#tagging-a-release)                              |

## Creating a release branch

A release branch should be created four weeks before the planned bpftrace
release. From that moment, only relevant bug fixes should be backported to the

> branch. Similarly, after the release is done, backport PRs should be opened

Reviewer: Not sure what the last sentence means.

Author: It wants to say that future bug fixes that should get into that minor release (by creating a new "patch" release) should be backported from master to this branch. It's actually taken from the [original document](https://github.com/bpftrace/bpftrace/blob/master/docs/release_process.md?plain=1#L24) but I agree that it would be nice to reformulate to make it clearer.

Reviewer: Yeah, just copy and paste this reply 😄

Author: I ended up moving this entire sentence (purpose of the branch) to the above section.

against this branch.

> The purpose of the branch is to give sufficient time to test features in the

Reviewer: ```suggestion
The purpose of this release branch is to give sufficient time to test features in the
```

> upcoming rebase without blocking the development on the master branch.

Reviewer: "upcoming rebase" - do you mean the new LLVM version?

Author: Nope, it's a typo, should be "upcoming release" meaning the new bpftrace release. Will fix.

When creating a branch, the following steps should be performed:
1. Create a new branch according to the [Branching model](#branching-model).
1. Update Nixpks to the latest version to get the latest (pre-release) LLVM by
   running
   ```
   nix flake update
   ```
   and committing the `flake.lock` changes to the repo. At this time, the `-rc2`
   or `-rc3` version of LLVM should be available.

> 1. Bump the supported LLVM version in `CMakeLists.txt` and `flake.nix`, resolve

Reviewer: Super nit but CmakeLists.txt and flaxe.nix aren't code - so no '`' - also you could have links to these two files here.

Reviewer: Same for '.github/workflows/ci.yml' below

Author: Good idea, will add the links.

Author: Also added links for file references from the original doc version (last section).

   any potential issues, and add a CI job to `.github/workflows/ci.yml` for the
   new version.
1. Once the final LLVM is released and present in Nixpkgs (usually 2-5 days
   after the LLVM release), repeat step 2 to get the released LLVM in the CI
   environment.

> Steps 2-4 should be done in the master branch first and then backported to the

Reviewer: This seems like an after thought, maybe it should be inline with the steps or above them.

Author: Done, probably worth checking again.

> release branch. In the rare case when master-first approach is not possible

Reviewer: ```suggestion
release branch. In the rare case when the master-first approach is not possible
```

| release branch created | **2 weeks before the LLVM release** | [Creating a release branch](#creating-a-release-branch)              |
| LLVM release           | usually second Tuesday of Mar/Sep   | [LLVM release schedule](https://llvm.org/docs/HowToReleaseLLVM.html) |
| bpftrace release       | **2 weeks after the LLVM release**  | [Tagging a release](#tagging-a-release)                              |

## Creating a release branch

A release branch should be created four weeks before the planned bpftrace
release. From that moment, only relevant bug fixes should be backported to the
branch.

The purpose of this release branch is to give sufficient time to test features
in the upcoming bpftrace release without blocking the development on the master
branch.

When creating a branch, the following steps should be performed. Any changes to
the code should be done in the master branch first and then backported to the
release branch. In the rare case when the master-first approach is not possible
(e.g. a feature present exclusively on master blocks the LLVM update), the

> changes can be done in the release branch first and backported to master

Reviewer 2: ```suggestion
changes can be done in the release branch first and forward-ported to master
```

Author: Good idea, fixed.

    * release/1.0.x
    * release/1.1.x

## Tagging a release

You must do the following steps to formally release a version.

In the release branch:

1. Mark the release in the CHANGELOG by replacing the `## Unreleased` header
   with `## [VERSION] date`.
1. Update `bpftrace_VERSION_MAJOR`, `bpftrace_VERSION_MINOR`, and
   `bpftrace_VERSION_PATCH` in `CMakeLists.txt` to the target version.
1. Tag a release. We do this in the github UI by clicking "releases" (on same
   line as "commits"), then "Draft a new release". The tag version and release
   title should be the same and in `vX.Y.Z` format. The tag description should
   be the same as what you added to `CHANGELOG.md`.
1. Check that automation picks up the new release and uploads release assets to
   the release.
1. If automation fails, please fix the automation for next time and also
   manually build+upload artifacts by running `scripts/create-assets.sh` from
   bpftrace root dir and attach the generated archives to the release.

Once the release is out:
1. Backport the `CHANGELOG`, `CMakeLists.txt`, and `flake.lock` changes from the

>    release branch to master.

Reviewer 2: Why does only the release branch get the new LLVM? Wouldn't we want master to also have it during the release stabilization window?

Author: Good point. My goal was to avoid updating LLVM in master before all errors are fixed which would block the development but it makes sense to backport it once the new CI job passes on the release branch. Let me add that.

    * release/1.0.x
    * release/1.1.x

## Tagging a release

You must do the following steps to formally release a version.

In the release branch:

1. Mark the release in [CHANGELOG.md](../CHANGELOG.md) by replacing the `##
   Unreleased` header with `## [VERSION] date`.
1. Update `bpftrace_VERSION_MAJOR`, `bpftrace_VERSION_MINOR`, and
   `bpftrace_VERSION_PATCH` in [CMakeLists.txt](../CMakeLists.txt) to the target
   version.
1. Tag a release. We do this in the github UI by clicking "releases" (on same
   line as "commits"), then "Draft a new release". The tag version and release
   title should be the same and in `vX.Y.Z` format. The tag description should
   be the same as what you added to CHANGELOG.md.
1. Check that automation picks up the new release and uploads release assets to
   the release.
1. If automation fails, please fix the automation for next time and also
   manually build+upload artifacts by running `scripts/create-assets.sh` from
   bpftrace root dir and attach the generated archives to the release.

Once the release is out:

> 1. Backport the CHANGELOG.md changes from the release branch to master.

Reviewer 2: ```suggestion
1. Forward-port the CHANGELOG.md changes from the release branch to master.
```

Author: Fixed
